bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/ColumnViewer.java

```diff
@@ -817,4 +841,194 @@ protected void setBusy(boolean busy) {
 	public boolean isBusy() {
 		return busy;
 	}
+
+	@Override
+	protected Object[] getSortedChildren(Object parent) {
+		Object[] sorted = super.getSortedChildren(parent);
+		return applyItemsLimit(parent, sorted);
+	}
+
+	/**
+	 * {@link StructuredViewer#getSortedChildren(Object)} always returns the
+	 * complete list of elements to be populated. Now we have to apply viewer limit
+	 * for the items to be created on the viewer. Note that in case of parent is
+	 * {@link IExpandableNode} we will return next block of limited elements to be
+	 * created.
+	 *
+	 * @param parent
+	 * @param sorted
+	 * @return returns only limited items.
+	 * @since 3.31
+	 */
+	protected Object[] applyItemsLimit(Object parent, Object[] sorted) {
+		// limit the number of items to be created. sorted always gets the remaining
+		// elements to be created.
+		int itemsLimit = getItemsLimit();
+		if (itemsLimit <= 0) {
+			return sorted;
+		}
+
+		if (sorted.length <= itemsLimit) {
+			return sorted;
+		}
+
+		int offSet = itemsLimit;
+		int srcPos = 0;
+
+		if (parent instanceof IExpandableNode expNode) {
+			srcPos = expNode.getOffSet();
+			offSet = srcPos + itemsLimit;
+			// pass on original children
+			sorted = expNode.getAllElements();
+		}
+
+		Object[] partialChildren = new Object[itemsLimit + 1];
+
+		// Extract a subset of children
+		System.arraycopy(sorted, srcPos, partialChildren, 0, itemsLimit);
+
+		// Add an expandable node
+		partialChildren[itemsLimit] = createExpandableNode(sorted, offSet, getItemsLimit());
+
+		return partialChildren;
+	}
+
+	/**
+	 * Viewer must dispose this cell and populate the remaining elements returned by
+	 * this cell.
+	 *
+	 * @param cell
+	 * @since 3.31
+	 */
+	protected void handleExpandableNodeClicked(Widget cell) {
+		// default implementation does nothing. Actual viewers can decide how to
+		// populate remaining elements.
+	}
+
+	/**
+	 * We should not dispose already visible items and we should display limited
+	 * elements along with already visible items.
+	 *
+	 * @param parent          model item to be refreshed.
+	 * @param visibleChildren currently visible children. This includes any elements
+	 *                        already expanded by user.
+	 * @return list of children to be displayed/refreshed.
+	 * @since 3.31
+	 */
+	protected Object[] getVisibleLimitBasedChildren(Object parent, Item[] visibleChildren) {
+		Object[] updatedChildren = null;
+		final int limit = getItemsLimit();
+		// at least one iteration of expandable items are populated.
+		if (visibleChildren.length > limit) {
+			List<Object> listOfEle = new ArrayList<>(Arrays.asList(getSortedChildren(parent)));
```

Comment:
this array -> list -> array conversion here and below can be avoided. It costs time/memory especially if we operate on elements number way over limit.

Reply:
@raghucssit : please double check my changes in this method.

Reply:
This change looks good..
Thank you..